File: results/dnd-rest-benchmark/lifecycle-runs/20260711T035512Z_pi_glm-5p2_python-django/dndsite/storage.py
"""SQLite-backed durable storage for the D&D REST API.

This module owns the single SQLite connection used by the whole process and
exposes a tiny data-access API for the durable game-state (combat sessions and
auth users). Schema is initialized on import (i.e. on server startup) so the
database file ``game.db`` exists before the first request is served.
"""
from __future__ import annotations

import json
import os
import sqlite3
import threading

# Resolve ``game.db`` in the project directory (the parent of this package).
_PROJECT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DB_PATH = os.path.join(_PROJECT_DIR, "game.db")

SCHEMA_VERSION = 1

_LOCK = threading.Lock()
_CONN: sqlite3.Connection | None = None


def _connect() -> sqlite3.Connection:
    conn = sqlite3.connect(DB_PATH, check_same_thread=False)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")
    return conn


_SCHEMA_SQL = """
CREATE TABLE IF NOT EXISTS meta (
    key   TEXT PRIMARY KEY,
    value TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS users (
    username      TEXT PRIMARY KEY,
    role          TEXT NOT NULL,
    password_hash TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS combat_sessions (
    id              TEXT PRIMARY KEY,
    round           INTEGER NOT NULL,
    turn_index      INTEGER NOT NULL,
    order_json      TEXT NOT NULL,
    conditions_json TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS monsters (
    slug        TEXT PRIMARY KEY,
    name        TEXT NOT NULL,
    cr          TEXT NOT NULL,
    armor_class INTEGER NOT NULL,
    hit_points  INTEGER NOT NULL,
    tags_json   TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS items (
    slug    TEXT PRIMARY KEY,
    name    TEXT NOT NULL,
    type    TEXT NOT NULL,
    rarity  TEXT NOT NULL,
    cost_gp INTEGER NOT NULL
);
CREATE TABLE IF NOT EXISTS campaigns (
    id   TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    dm   TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS campaign_characters (
    campaign_id TEXT NOT NULL,
    id          TEXT NOT NULL,
    name        TEXT NOT NULL,
    level       INTEGER NOT NULL,
    "class"     TEXT NOT NULL,
    PRIMARY KEY (campaign_id, id),
    FOREIGN KEY (campaign_id) REFERENCES campaigns(id) ON DELETE CASCADE
);
CREATE TABLE IF NOT EXISTS campaign_events (
    campaign_id TEXT NOT NULL,
    id          TEXT NOT NULL,
    kind        TEXT NOT NULL,
    summary     TEXT NOT NULL DEFAULT '',
    PRIMARY KEY (campaign_id, id),
    FOREIGN KEY (campaign_id) REFERENCES campaigns(id) ON DELETE CASCADE
);
"""


def init_schema(reset: bool = False) -> None:
    """Create the database file and tables.

    With ``reset=True`` all benchmark-created durable data is dropped first and
    the schema is recreated from scratch.
    """
    global _CONN
    with _LOCK:
        if _CONN is None:
            _CONN = _connect()
        cur = _CONN.cursor()
        if reset:
            cur.executescript(
                "DROP TABLE IF EXISTS campaign_events;"
                "DROP TABLE IF EXISTS campaign_characters;"
                "DROP TABLE IF EXISTS campaigns;"
                "DROP TABLE IF EXISTS users;"
                "DROP TABLE IF EXISTS combat_sessions;"
                "DROP TABLE IF EXISTS monsters;"
                "DROP TABLE IF EXISTS items;"
                "DROP TABLE IF EXISTS meta;"
            )
        cur.executescript(_SCHEMA_SQL)
        cur.execute(
            "INSERT OR REPLACE INTO meta(key, value) VALUES (?, ?)",
            ("schema_version", str(SCHEMA_VERSION)),
        )
        _CONN.commit()


def is_initialized() -> bool:
    """Return True when the schema has been created in the database file."""
    with _LOCK:
        if _CONN is None:
            return False
        cur = _CONN.cursor()
        cur.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name='meta'"
        )
        if cur.fetchone() is None:
            return False
        cur.execute("SELECT value FROM meta WHERE key='schema_version'")
        row = cur.fetchone()
        if row is None:
            return False
        try:
            return int(row["value"]) == SCHEMA_VERSION
        except (TypeError, ValueError):
            return False


def reset() -> None:
    """Drop and recreate all benchmark-created durable data."""
    init_schema(reset=True)


# ---------------------------------------------------------------------------
# Users
# ---------------------------------------------------------------------------

def get_user(username: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT username, role, password_hash FROM users WHERE username=?",
            (username,),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "username": row["username"],
            "role": row["role"],
            "password_hash": row["password_hash"],
        }


def insert_user(username: str, role: str, password_hash: str) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "INSERT INTO users(username, role, password_hash) VALUES (?, ?, ?)",
            (username, role, password_hash),
        )
        _CONN.commit()


# ---------------------------------------------------------------------------
# Combat sessions
# ---------------------------------------------------------------------------

def upsert_session(session: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            """
            INSERT INTO combat_sessions(id, round, turn_index, order_json, conditions_json)
            VALUES (?, ?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET
                round           = excluded.round,
                turn_index      = excluded.turn_index,
                order_json      = excluded.order_json,
                conditions_json = excluded.conditions_json
            """,
            (
                session["id"],
                session["round"],
                session["turn_index"],
                json.dumps(session["order"]),
                json.dumps(session["conditions"]),
            ),
        )
        _CONN.commit()


def get_session(sid: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT id, round, turn_index, order_json, conditions_json "
            "FROM combat_sessions WHERE id=?",
            (sid,),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "id": row["id"],
            "round": row["round"],
            "turn_index": row["turn_index"],
            "order": json.loads(row["order_json"]),
            "conditions": json.loads(row["conditions_json"]),
        }


def delete_session(sid: str) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute("DELETE FROM combat_sessions WHERE id=?", (sid,))
        _CONN.commit()


# ---------------------------------------------------------------------------
# Compendium: monsters and items
# ---------------------------------------------------------------------------

def insert_monster(monster: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "INSERT INTO monsters(slug, name, cr, armor_class, hit_points, tags_json) "
            "VALUES (?, ?, ?, ?, ?, ?)",
            (
                monster["slug"],
                monster["name"],
                monster["cr"],
                monster["armor_class"],
                monster["hit_points"],
                json.dumps(monster["tags"]),
            ),
        )
        _CONN.commit()


def get_monster(slug: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT slug, name, cr, armor_class, hit_points, tags_json "
            "FROM monsters WHERE slug=?",
            (slug,),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "slug": row["slug"],
            "name": row["name"],
            "cr": row["cr"],
            "armor_class": row["armor_class"],
            "hit_points": row["hit_points"],
            "tags": json.loads(row["tags_json"]),
        }


def insert_item(item: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "INSERT INTO items(slug, name, type, rarity, cost_gp) "
            "VALUES (?, ?, ?, ?, ?)",
            (
                item["slug"],
                item["name"],
                item["type"],
                item["rarity"],
                item["cost_gp"],
            ),
        )
        _CONN.commit()


def get_item(slug: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug=?",
            (slug,),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "slug": row["slug"],
            "name": row["name"],
            "type": row["type"],
            "rarity": row["rarity"],
            "cost_gp": row["cost_gp"],
        }


# ---------------------------------------------------------------------------
# Campaign state: campaigns, characters, and session log events
# ---------------------------------------------------------------------------

def insert_campaign(campaign: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "INSERT INTO campaigns(id, name, dm) VALUES (?, ?, ?)",
            (campaign["id"], campaign["name"], campaign["dm"]),
        )
        _CONN.commit()


def get_campaign(cid: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT id, name, dm FROM campaigns WHERE id=?",
            (cid,),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "id": row["id"],
            "name": row["name"],
            "dm": row["dm"],
        }


def get_character(campaign_id: str, char_id: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            'SELECT id, name, level, "class" FROM campaign_characters '
            "WHERE campaign_id=? AND id=?",
            (campaign_id, char_id),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "id": row["id"],
            "name": row["name"],
            "level": row["level"],
            "class": row["class"],
        }


def insert_character(campaign_id: str, character: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            'INSERT INTO campaign_characters(campaign_id, id, name, level, "class") '
            "VALUES (?, ?, ?, ?, ?)",
            (
                campaign_id,
                character["id"],
                character["name"],
                character["level"],
                character["class"],
            ),
        )
        _CONN.commit()


def list_characters(campaign_id: str) -> list[dict]:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            'SELECT id, name, level, "class" FROM campaign_characters '
            "WHERE campaign_id=? ORDER BY rowid",
            (campaign_id,),
        )
        return [
            {
                "id": row["id"],
                "name": row["name"],
                "level": row["level"],
                "class": row["class"],
            }
            for row in cur.fetchall()
        ]


def get_event(campaign_id: str, evt_id: str) -> dict | None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT id, kind, summary FROM campaign_events "
            "WHERE campaign_id=? AND id=?",
            (campaign_id, evt_id),
        )
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "id": row["id"],
            "kind": row["kind"],
            "summary": row["summary"],
        }


def insert_event(campaign_id: str, event: dict) -> None:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "INSERT INTO campaign_events(campaign_id, id, kind, summary) "
            "VALUES (?, ?, ?, ?)",
            (
                campaign_id,
                event["id"],
                event["kind"],
                event["summary"],
            ),
        )
        _CONN.commit()


def count_events(campaign_id: str) -> int:
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT COUNT(*) AS n FROM campaign_events WHERE campaign_id=?",
            (campaign_id,),
        )
        row = cur.fetchone()
        return row["n"] if row is not None else 0


def list_events(campaign_id: str) -> list[dict]:
    """Return the campaign's logged events ordered by insertion sequence."""
    with _LOCK:
        cur = _CONN.cursor()
        cur.execute(
            "SELECT id, kind, summary FROM campaign_events "
            "WHERE campaign_id=? ORDER BY rowid",
            (campaign_id,),
        )
        return [
            {"id": row["id"], "kind": row["kind"], "summary": row["summary"]}
            for row in cur.fetchall()
        ]


# Initialize the schema as soon as the module is imported (server startup).
init_schema()

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081339Z_claude_claude-opus-5_go-open/storage.go
package main

import (
	"database/sql"
	"log"
	"net/http"
	"strings"
	"sync"

	_ "modernc.org/sqlite"
)

// Durable game-world and game-state data lives in a SQLite database file next
// to the binary. Every mutation goes through this file so a restart finds the
// schema already in place.

const (
	storageDriver = "sqlite"
	schemaVersion = 1
	databaseFile  = "game.db"
)

// db is the single shared handle. Connections are capped at one so writers
// serialize inside SQLite instead of racing for the write lock, which keeps
// behavior deterministic under the evaluator's concurrent requests.
var db *sql.DB

// storeMu guards read-modify-write sequences that span multiple statements
// (session creation, condition ticking) so they stay atomic end to end.
var storeMu sync.Mutex

const schemaDDL = `
CREATE TABLE IF NOT EXISTS schema_meta (
	id INTEGER PRIMARY KEY CHECK (id = 1),
	version INTEGER NOT NULL
);
CREATE TABLE IF NOT EXISTS users (
	username TEXT PRIMARY KEY,
	role     TEXT NOT NULL,
	salt     BLOB NOT NULL,
	digest   BLOB NOT NULL
);
CREATE TABLE IF NOT EXISTS combat_sessions (
	id         TEXT PRIMARY KEY,
	round      INTEGER NOT NULL,
	turn_index INTEGER NOT NULL
);
CREATE TABLE IF NOT EXISTS combat_combatants (
	session_id TEXT NOT NULL REFERENCES combat_sessions(id) ON DELETE CASCADE,
	position   INTEGER NOT NULL,
	name       TEXT NOT NULL,
	dex        INTEGER NOT NULL,
	score      INTEGER NOT NULL,
	PRIMARY KEY (session_id, position)
);
-- A row with condition IS NULL marks a combatant that has held a condition at
-- some point, so an emptied list still renders as [] rather than disappearing.
CREATE TABLE IF NOT EXISTS combat_conditions (
	session_id TEXT NOT NULL REFERENCES combat_sessions(id) ON DELETE CASCADE,
	target     TEXT NOT NULL,
	position   INTEGER NOT NULL,
	condition  TEXT,
	remaining  INTEGER
);
CREATE TABLE IF NOT EXISTS monsters (
	slug        TEXT PRIMARY KEY,
	name        TEXT NOT NULL,
	cr          TEXT NOT NULL,
	armor_class INTEGER NOT NULL,
	hit_points  INTEGER NOT NULL,
	tags        TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS items (
	slug    TEXT PRIMARY KEY,
	name    TEXT NOT NULL,
	type    TEXT NOT NULL,
	rarity  TEXT NOT NULL,
	cost_gp REAL NOT NULL
);
CREATE TABLE IF NOT EXISTS campaigns (
	id   TEXT PRIMARY KEY,
	name TEXT NOT NULL,
	dm   TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS campaign_characters (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id          TEXT NOT NULL,
	position    INTEGER NOT NULL,
	name        TEXT NOT NULL,
	level       INTEGER NOT NULL,
	class       TEXT NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
CREATE TABLE IF NOT EXISTS campaign_events (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id          TEXT NOT NULL,
	position    INTEGER NOT NULL,
	kind        TEXT NOT NULL,
	summary     TEXT NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
CREATE TABLE IF NOT EXISTS campaign_quests (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id          TEXT NOT NULL,
	position    INTEGER NOT NULL,
	title       TEXT NOT NULL,
	status      TEXT NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
CREATE TABLE IF NOT EXISTS campaign_quest_milestones (
	campaign_id TEXT NOT NULL,
	quest_id    TEXT NOT NULL,
	position    INTEGER NOT NULL,
	name        TEXT NOT NULL,
	done        INTEGER NOT NULL,
	PRIMARY KEY (campaign_id, quest_id, name),
	FOREIGN KEY (campaign_id, quest_id)
		REFERENCES campaign_quests(campaign_id, id) ON DELETE CASCADE
);
CREATE TABLE IF NOT EXISTS campaign_factions (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id          TEXT NOT NULL,
	position    INTEGER NOT NULL,
	name        TEXT NOT NULL,
	stance      TEXT NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
-- faction_id is a plain column, not a foreign key: it is empty for an
-- unaffiliated NPC, and membership is checked by the handler instead.
CREATE TABLE IF NOT EXISTS campaign_npcs (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id          TEXT NOT NULL,
	position    INTEGER NOT NULL,
	name        TEXT NOT NULL,
	faction_id  TEXT NOT NULL,
	disposition INTEGER NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
-- Inventory and equipment rows are append-only stacks keyed by position, not by
-- slug: the same item added twice is two rows, and the summary counts rows.
CREATE TABLE IF NOT EXISTS campaign_inventory (
	campaign_id TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	position    INTEGER NOT NULL,
	item_slug   TEXT NOT NULL,
	quantity    INTEGER NOT NULL,
	owner       TEXT NOT NULL,
	PRIMARY KEY (campaign_id, position)
);
CREATE TABLE IF NOT EXISTS campaign_equipment (
	campaign_id  TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	position     INTEGER NOT NULL,
	character_id TEXT NOT NULL,
	item_slug    TEXT NOT NULL,
	quantity     INTEGER NOT NULL,
	PRIMARY KEY (campaign_id, position)
);
-- A downtime crafting project accrues days_completed until it reaches
-- days_required, at which point its status flips to complete. It is a record of
-- its own and does not write to campaign_inventory.
CREATE TABLE IF NOT EXISTS campaign_crafting (
	campaign_id    TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id             TEXT NOT NULL,
	position       INTEGER NOT NULL,
	character_id   TEXT NOT NULL,
	item_slug      TEXT NOT NULL,
	days_required  INTEGER NOT NULL,
	days_completed INTEGER NOT NULL,
	cost_gp        REAL NOT NULL,
	status         TEXT NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
-- A scheduled play session keeps the caller's starts_at verbatim for echoing
-- plus a normalized UTC key so ordering does not depend on how the timestamp
-- was spelled. Agenda and attendance are its ordered/keyed children.
CREATE TABLE IF NOT EXISTS campaign_sessions (
	campaign_id      TEXT NOT NULL REFERENCES campaigns(id) ON DELETE CASCADE,
	id               TEXT NOT NULL,
	position         INTEGER NOT NULL,
	starts_at        TEXT NOT NULL,
	starts_at_key    TEXT NOT NULL,
	duration_minutes INTEGER NOT NULL,
	PRIMARY KEY (campaign_id, id)
);
CREATE TABLE IF NOT EXISTS campaign_session_agenda (
	campaign_id TEXT NOT NULL,
	session_id  TEXT NOT NULL,
	position    INTEGER NOT NULL,
	entry       TEXT NOT NULL,
	PRIMARY KEY (campaign_id, session_id, position),
	FOREIGN KEY (campaign_id, session_id)
		REFERENCES campaign_sessions(campaign_id, id) ON DELETE CASCADE
);
CREATE TABLE IF NOT EXISTS campaign_session_attendance (
	campaign_id  TEXT NOT NULL,
	session_id   TEXT NOT NULL,
	character_id TEXT NOT NULL,
	status       TEXT NOT NULL,
	PRIMARY KEY (campaign_id, session_id, character_id),
	FOREIGN KEY (campaign_id, session_id)
		REFERENCES campaign_sessions(campaign_id, id) ON DELETE CASCADE
);
-- Play campaigns are the authenticated /v1/play surface and are deliberately
-- separate from the open campaigns table above: they carry an owner (the dm
-- who created them) and a lifecycle status, and ids never collide across the two.
-- current_actor and turn_number are the play clock. They stay empty/0 while the
-- campaign sits in the lobby and are set once, when the owner starts it.
-- turn_started_seq records the event-log position the current turn opened at,
-- which is what the turn timeout policy measures against; nudge_count is the
-- running total of owner nudges and never decreases.
CREATE TABLE IF NOT EXISTS play_campaigns (
	id               TEXT PRIMARY KEY,
	name             TEXT NOT NULL,
	owner            TEXT NOT NULL,
	status           TEXT NOT NULL,
	max_players      INTEGER NOT NULL,
	current_actor    TEXT NOT NULL DEFAULT '',
	turn_number      INTEGER NOT NULL DEFAULT 0,
	turn_started_seq INTEGER NOT NULL DEFAULT 0,
	nudge_count      INTEGER NOT NULL DEFAULT 0
);
-- One party membership per (campaign, player). The character id is unique
-- within the campaign too, and position records join order so the party always
-- renders in the sequence players actually joined.
CREATE TABLE IF NOT EXISTS play_campaign_members (
	campaign_id  TEXT NOT NULL REFERENCES play_campaigns(id) ON DELETE CASCADE,
	username     TEXT NOT NULL,
	position     INTEGER NOT NULL,
	character_id TEXT NOT NULL,
	name         TEXT NOT NULL,
	class        TEXT NOT NULL,
	PRIMARY KEY (campaign_id, username),
	UNIQUE (campaign_id, character_id)
);
-- The play event log is append-only: sequence starts at 1 per campaign and is
-- never rewritten, so the story reads back in the order it was told.
CREATE TABLE IF NOT EXISTS play_campaign_events (
	campaign_id TEXT NOT NULL REFERENCES play_campaigns(id) ON DELETE CASCADE,
	sequence    INTEGER NOT NULL,
	kind        TEXT NOT NULL,
	actor       TEXT NOT NULL,
	text        TEXT NOT NULL,
	action_type TEXT NOT NULL DEFAULT '',
	PRIMARY KEY (campaign_id, sequence)
);
INSERT INTO schema_meta (id, version) VALUES (1, 1)
	ON CONFLICT(id) DO UPDATE SET version = excluded.version;
`

// openStorage opens the database and installs the schema. busy_timeout keeps a
// contended write waiting instead of failing outright; foreign_keys makes the
// ON DELETE CASCADE declarations in the schema actually fire.
func openStorage() error {
	handle, err := sql.Open(storageDriver, "file:"+databaseFile+"?_pragma=busy_timeout(5000)&_pragma=foreign_keys(1)")
	if err != nil {
		return err
	}
	handle.SetMaxOpenConns(1)
	db = handle
	return initSchema()
}

// initSchema is idempotent: every statement in schemaDDL is IF NOT EXISTS or an
// upsert, so it doubles as both first-run creation and post-reset recreation.
func initSchema() error {
	if _, err := db.Exec(schemaDDL); err != nil {
		return err
	}
	return migrateSchema()
}

// migrateSchema brings a database file created by an earlier build up to the
// current shape. CREATE TABLE IF NOT EXISTS leaves an existing table alone, so
// columns added after that table was first written have to be added here. Each
// step is idempotent: a column that is already present makes ALTER TABLE fail,
// and that failure is the signal there is nothing to do.
func migrateSchema() error {
	for _, stmt := range []string{
		`ALTER TABLE play_campaign_events ADD COLUMN action_type TEXT NOT NULL DEFAULT ''`,
		`ALTER TABLE play_campaigns ADD COLUMN turn_started_seq INTEGER NOT NULL DEFAULT 0`,
		`ALTER TABLE play_campaigns ADD COLUMN nudge_count INTEGER NOT NULL DEFAULT 0`,
	} {
		if _, err := db.Exec(stmt); err != nil && !strings.Contains(err.Error(), "duplicate column name") {
			return err
		}
	}
	return nil
}

// resetStorage drops every table and recreates the schema, giving each
// evaluation shot a clean slate. The process keeps running against the same
// handle throughout. Tables drop children first so foreign keys stay satisfied.
func resetStorage() error {
	storeMu.Lock()
	defer storeMu.Unlock()
	const dropDDL = `
DROP TABLE IF EXISTS play_campaign_events;
DROP TABLE IF EXISTS play_campaign_members;
DROP TABLE IF EXISTS play_campaigns;
DROP TABLE IF EXISTS campaign_session_attendance;
DROP TABLE IF EXISTS campaign_session_agenda;
DROP TABLE IF EXISTS campaign_sessions;
DROP TABLE IF EXISTS campaign_crafting;
DROP TABLE IF EXISTS campaign_equipment;
DROP TABLE IF EXISTS campaign_inventory;
DROP TABLE IF EXISTS campaign_npcs;
DROP TABLE IF EXISTS campaign_factions;
DROP TABLE IF EXISTS campaign_quest_milestones;
DROP TABLE IF EXISTS campaign_quests;
DROP TABLE IF EXISTS campaign_events;
DROP TABLE IF EXISTS campaign_characters;
DROP TABLE IF EXISTS campaigns;
DROP TABLE IF EXISTS combat_conditions;
DROP TABLE IF EXISTS combat_combatants;
DROP TABLE IF EXISTS combat_sessions;
DROP TABLE IF EXISTS users;
DROP TABLE IF EXISTS monsters;
DROP TABLE IF EXISTS items;
DROP TABLE IF EXISTS schema_meta;
`
	if _, err := db.Exec(dropDDL); err != nil {
		return err
	}
	return initSchema()
}

// storageInitialized reports whether the schema is present and stamped.
func storageInitialized() bool {
	var version int
	err := db.QueryRow(`SELECT version FROM schema_meta WHERE id = 1`).Scan(&version)
	return err == nil && version == schemaVersion
}

// ---------- GET /v1/storage/status ----------

type storageStatusResponse struct {
	Driver        string `json:"driver"`
	SchemaVersion int    `json:"schema_version"`
	Initialized   bool   `json:"initialized"`
}

func handleStorageStatus(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodGet) {
		return
	}
	writeJSON(w, http.StatusOK, storageStatusResponse{
		Driver:        storageDriver,
		SchemaVersion: schemaVersion,
		Initialized:   storageInitialized(),
	})
}

// ---------- POST /v1/storage/reset ----------

type storageResetResponse struct {
	OK            bool `json:"ok"`
	SchemaVersion int  `json:"schema_version"`
}

// handleStorageReset takes no body, so it guards its method directly rather
// than through decodeBody. Its 500 message is endpoint-specific, not one of the
// generic storage messages in httpx.go.
func handleStorageReset(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	if err := resetStorage(); err != nil {
		log.Printf("storage reset failed: %v", err)
		writeError(w, http.StatusInternalServerError, "storage reset failed")
		return
	}
	writeJSON(w, http.StatusOK, storageResetResponse{OK: true, SchemaVersion: schemaVersion})
}
